src/Pate/Equivalence/MemoryDomain.hs

```diff
@@ -0,0 +1,224 @@
+traverseWithCell memDom f = join $ traverseWithCellPar memDom (\cell p -> return $ f cell p)
+
+
+toList ::
+  MemoryDomain sym arch ->
+  [(Some (PMC.MemCell sym arch), W4.Pred sym)]
+toList memDom = M.toList (memDomainPred memDom)
+
+toAscList ::
+  MemoryDomain sym arch ->
+  [(Some (PMC.MemCell sym arch), W4.Pred sym)]
+toAscList memDom = M.toAscList (memDomainPred memDom)
+
+cells ::
+  OrdF (W4.SymExpr sym) =>
+  MemoryDomain sym arch ->
+  Set (Some (PMC.MemCell sym arch))
+cells memDom = S.fromList $ map fst (toList memDom)
+
+-- | Build a 'MemoryDomain' from a list of 'PMC.MemCell' entries, with a corresponding
+-- predicate indicating whether or not they are in or out of the domain (subject to the
+-- given polarity).
+-- Duplicate entries are allowed, where the resulting predicate will be the disjunction
+-- of the overlapping predicates.
+fromList ::
+  W4.IsExprBuilder sym =>
+  OrdF (W4.SymExpr sym) =>
+  sym ->
+  [(Some (PMC.MemCell sym arch), W4.Pred sym)] ->
+  W4.Pred sym ->
+  IO (MemoryDomain sym arch)
+fromList sym l pol = do
+  -- NOTE: We can't just use Data.Map.fromList here because it will discard duplicate
+  -- entries
+  let maps = map (\(cell, p) -> M.singleton cell p) l
+  merged <- foldM (PMC.mergeMemCellPred sym) M.empty maps
+  return $ MemoryDomain merged pol
+
+-- | Similar to 'fromList' but assumes the cells are in order and disjoint
+fromAscList ::
+  W4.IsExprBuilder sym =>
+  OrdF (W4.SymExpr sym) =>
+  [(Some (PMC.MemCell sym arch), W4.Pred sym)] ->
+  W4.Pred sym ->
+  MemoryDomain sym arch
+fromAscList l pol = MemoryDomain (M.fromAscList l) pol
+
+mux ::
+  W4.IsExprBuilder sym =>
+  OrdF (W4.SymExpr sym) =>
+  sym ->
+  W4.Pred sym ->
+  MemoryDomain sym arch ->
+  MemoryDomain sym arch ->
+  IO (MemoryDomain sym arch)
+mux sym p predT predF = case W4.asConstantPred p of
+  Just True -> return predT
+  Just False -> return predF
+  _ -> do
+    pol <- W4.baseTypeIte sym p (memDomainPolarity predT) (memDomainPolarity predF)
+    locs <- PMC.muxMemCellPred sym p (memDomainPred predT) (memDomainPred predF)
+    return $ MemoryDomain locs pol
+
+-- | True if the given 'PMC.MemCell' is in the given 'MemoryDomain', according to
+-- its polarity.
+containsCell ::
+  W4.IsExprBuilder sym =>
+  OrdF (W4.SymExpr sym) =>
+  sym ->
+  MemoryDomain sym arch ->
+  PMC.MemCell sym arch w ->
+  IO (W4.Pred sym)
+containsCell sym memDom cell = do
+  isInLocs <- PMC.inMemCellPred sym cell (memDomainPred memDom)
+  W4.isEq sym isInLocs (memDomainPolarity memDom)
+
+
+-- | Trivial domain that covers all of memory
+universal :: W4.IsExprBuilder sym => sym -> MemoryDomain sym arch
+universal sym = MemoryDomain M.empty (W4.falsePred sym)
+
+-- | Trivial domain that covers no memory (the empty domain)
+empty :: W4.IsExprBuilder sym => sym -> MemoryDomain sym arch
+empty sym = MemoryDomain M.empty (W4.truePred sym)
+
+
+fromFootPrints ::
```

Comment:
What is the concept for this function? There is some non-trivial logic around memory operation directions that probably merits discussion.

What expectations should the reader have for the resulting `MemoryDomain`?